eval/adjust.rkt

```diff
@@ -71,333 +71,263 @@
+           (+ (logspan x) (* 2 (logspan y))))]        ; exponent per y
+
+    [(ival-sqrt ival-cbrt)
+     ; sqrt: logspan(x)/2 - 1
+     ; cbrt: logspan(x)*2/3 - 1
+     (define x (first srcs))
+     (list (quotient (logspan x) 2))]
     
     [(ival-add ival-sub)
-     ; log[Г+]'x = log[x] - log[x + y] + 1 (1 for mantissa approximation when dividing)
-     ; log[Г+]'y = log[y] - log[x + y] + 1
-     ; log[Г-]'x = log[x] - log[x - y] + 1
-     ; log[Г-]'y = log[y] - log[x - y] + 1
-     ;                    ^^^^^^^^^^^^
-     ;                    possible cancellation
+     ; k = 1: maxlog(x) - minlog(z)
+     ; k = 2: maxlog(y) - minlog(z)
      (define x (first srcs))
      (define y (second srcs))
-     (define outlo (ival-lo output))
-     (define outhi (ival-hi output))
-     
-     (define x-exp (ival-max-log2-approx x))
-     (define y-exp (ival-max-log2-approx y))
-     (define out-exp (ival-min-log2-approx output))
      
-     (define slack (if (equal? (mpfr-sign outlo) (mpfr-sign outhi))
-                       0
-                       (get-slack)))
-     
-     (list (+ (- x-exp out-exp) slack 1)              ; exponent per x
-           (+ (- y-exp out-exp) slack 1))]            ; exponent per y 
+     (list (- (maxlog x) (minlog z))                  ; exponent per x
+           (- (maxlog y) (minlog z)))]                ; exponent per y 
     
     [(ival-pow)
-     ; log[Гpow]'x = log[y]
-     ; log[Гpow]'y = log[y] + log[log(x)] <= log[y] + |log[x]|
-
+     ; k = 1: maxlog(y) + logspan(x) + logspan(z)
+     ; k = 2: maxlog(y) + max(|minlog(x)|,|maxlog(x)|) + logspan(z)
      (define x (first srcs))
      (define y (second srcs))
-     (define outlo (ival-lo output))
-     (define outhi (ival-hi output))
      
-     (define x-exp (ival-max-log2-approx x))
-     (define y-exp (ival-max-log2-approx y))
-     
-     ; when output crosses zero and x is negative - means that y was fractional and not fixed
+     ; when output crosses zero and x is negative - means that y was fractional and not fixed (specific of Rival)
      ; solution - add more slack for y to converge
-     (define y-slack (if (and (not (equal? (mpfr-sign outlo) (mpfr-sign outhi)))
-                              (bfnegative? (ival-lo x)))
-                         (get-slack)
-                         0))
-
-     #;(define xlo-exp (mpfr-exp (ival-lo x)))
-     #;(define xhi-exp (mpfr-exp (ival-hi x)))
-     #;(define x-slack ; it is likely to be a handling case from IEEE, x is not close enough
-       (if (or (equal? xlo-exp -9223372036854775807)
-               (equal? xhi-exp -9223372036854775807)        ; if interval contains 0
-               (and (equal? 1 xlo-exp) (equal? 1 xhi-exp))) ; if interval possibly contains 1
-           (get-slack)
-           0))
+     (define slack (if (and (crosses-zero? z) (bfnegative? (ival-lo x)))
+                       (get-slack)
+                       0))
```

Comment:
Why do we need the slack here? Something with negative bases? Fine, they're dumb

Reply:
Yeah, I added this case for `ival-pow-neg`. When the output crosses zero and `x` is negative it means that `y` was not fixed enough.
```
(define (ival-pow-neg x y)
  ;; Assumes x is negative
  (if (bf=? (ival-lo-val y) (ival-hi-val y))
      (if (bfinteger? (ival-lo-val y))
          ; If y is an integer point interval, there's no error,
          ; because it's always valid to raise to an integer power.
          (if (bfodd? (ival-lo-val y))
              (ival-neg (ival-pow-pos (ival-exact-fabs x) y)) ; Use fabs in case of [x, 0]
              (ival-pow-pos (ival-exact-fabs x) y))
          ; If y is non-integer point interval, it must be an even
          ; fraction (because all bigfloats are) so we always error
          ival-illegal)
      ; Moreover, if we have (-x)^y, that's basically x^y U -(x^y).
      (let ([pospow (ival-pow-pos (ival-exact-fabs x) y)])
        (ival-then (ival-assert ival-maybe) (ival-union (ival-neg pospow) pospow)))))
```
      
Slack helps here to hit the correct value faster